### ISSUE TITLE
Angular-mocks failing unit test as $destroy method is not a function

### ISSUE THREAD

**Issue by darkmavis1980** [angular/angular.js#13985]
Just tried to update angular on a project I'm working, and suddenly all the unit tests started to fail, simply because the `$destroy` method in **angular-mocks** is not a function anymore for some reasons.
That's a sample of the error I'm getting with Karma:
```
TypeError: injector.get(...).$destroy is not a function
                at Object.<anonymous> (/app/bower_components/angular-mocks/angular-mocks.js:2612:34)
```
I tried to take a look at the source code and I couldn't find any $destroy function, but it might be inherited by other scripts and for some reason is not anymore accessible to angular-mocks.

**Comment by gkalpak** [comment]
Can you create a live demo (e.g. using CodePen, Plnkr etc) ? It's impossible to tell what's going on without more info.

BTW, is batarang invovled by ay chance ?

**Comment by darkmavis1980** [comment]
I'll see if I can do it, but no, I'm not using batarang

**Comment by darkmavis1980** [comment]
Uhm not 100% sure how I can run unit tests with jasmine and karma on Codepen, but what I can see that my tests, if I use any version of 1.4.x runs fine, but if I switch to 1.5.0 a good 40% of them fails because of this ngMocks error

**Comment by gkalpak** [comment]
@darkmavis1980, you can't use karma in CodePen (but hopefully that's not contributing to the problem).
Here is a [sample pen][1] you can fork, featuring Angular+Jasmine.

[1]: https://codepen.io/ExpertSystem/pen/gbZVbb?editors=0011

**Comment by darkmavis1980** [comment]
Ok, I'll give it a try later to see if I can replicate the issue, I'll also double check if any of the libraries, such jasmine, karma and the various plugins are up to date

**Comment by darkmavis1980** [comment]
I did try to replicate it on CodePen, but there it seems to work fine, but there there are just a few libraries loaded, perhaps it's the mix with other angular libraries that creates the problem.
I've also tried to update all the dependencies we use on our project to the latest version, but nothing change, a good percentage of the test breaks with all the same issue.
I initially thought it was just on filters, but looking through the jasmine report, it's kinda random, and some test are quite trivial as well, nothing really complex.
Again, if I revert back to 1.4.9, everything works fine

**Comment by gkalpak** [comment]
This is strange. The line that this is failing on includes: `$injector.get('$rootScope').$destroy()`

Since this is failing because `$destroy` is not a function, it means that it can successfully retrieve `$rootScope`, but the `$rootScope`'s `$destroy` method is missing. This is definitely not done by Angular.

--
This might be an issue with some test (or some other library) is mocking out `$rootScope` or that `angular-mocks` doesn't clean up correctly.

From what you said earlier, I assume you are using jasmine (latest 2.x version ?) and karma and some other libs. Does any of the other libs fiddle with decorating or mocking `$rootScope` ? Are you using jasmine's `beforeAll` (I guess not).

Without a reproduction, it's unfortunately really hard to identify the problem.

Thing that might help:

1. Determining which lib (if any) is causing the conflict. E.g. trying to remove them one-by-one until the problem goes away. (I understand this might not be possible, due to the testing setup.)
2. Determining exactly which Angular version introduced the issue. You can try all the betas and RCs and see in which one does the problem start to appear. (Versions to try: `v1.5.0-beta.0`, `v1.5.0-beta.1`, `v1.5.0-beta.2`, `v1.5.0-rc.0`, `v1.5.0-rc.1`, `v1.5.0-rc.2`)

**Comment by darkmavis1980** [comment]
Yeah you are right, using jasmine 2.x and karma, plus various libraries.
The problem that it fails even where I don't use the $rootScope, like filters or some services.
I'll do further investigations and I'll try to figure out what's the cause, if you want you can close the bug for now as it seems to be a vary specific case, unless somebody else pop in the discussion.
Thanks for the help anyway, and I'll keep you posted if I find anything

**Comment by gkalpak** [comment]
It fails even if you don't use `$rootScope` explicitly in your tests, because `$rootScope` will always be created when bootstrapping an app and the `$injector` will always try to clean up.

If the reason it fails is that some lib messes with `$rootScope`, that's not Angular's fault, but if it fails because something leaks from previous test, then it's probably oru fault.
One way to find that out, would be to run all tests, find a simple one that fails while not using `$rootScope` and then run that test in isolation (i.e. disabled all other tests) and see if it fails or not.

BTW, are the same tests failing or is it random ?

--
Anyway, I'll keep this open for a few days in case something comes up. Thx for looking into this :smiley:

**Comment by darkmavis1980** [comment]
Yeah, the same tests are failing, but the randomness is the fact that some are filters (very simple ones) and some others are services or factories.
I'll try to investigate what's the source of the problem, I'll also try to test with all the angular 1.5 versions to see from which version it starts to appear.
I'll let you know!

**Comment by sunilkolla** [comment]
Try to update PhantomJS 2.*

**Comment by aciccarello** [comment]
I ran into this issue when upgrading from angular 1.4.8 to 1.5.0. The tests where it is failing for me are with a `module.run()` function. using `$rootScope.$on()` to register a listener on a ui-router event. My tests mocks `$rootScope` to capture functions registered with `$on()`.

After running into the issue, I tried to upgrade ui-router (0.2.15 -> 0.2.18) but that didn't change the results. I'll probably refactor my tests to get around the issue but I'm surprised that the error was caused by the upgrade.

**Comment by gkalpak** [comment]
@aciccarello, if you can provide a live demo or a minimal code sample that reproduces the problem, I'd be curious to take a look.

**Comment by aciccarello** [comment]
@gkalpak, I started putting together a plunker but I haven't gotten it working/not working with the respective versions of angular yet. I'll post it here if I get it working.

Essentially I was mocking the `$rootScope` service and after updating I started getting the error. The error is thrown from the afterEach function of angular-mocks.

```TypeScript
// Setup mocks
  beforeEach(angular.mock.module({
    $rootScope: { $on: (eventKey: string, fn) => {
        if (eventKey === '$stateChangeStart') {
          $stateChangeStart = fn;
        }
      }
    },
    $state: jasmine.createSpyObj('$state', ['includes', 'go']),
    $window: {}
  }));
```

**Comment by gkalpak** [comment]
@aciccarello, thx for the code snippet. Your problem should be reated to b75c0d8 (included in `v1.5.0-rc.0`): `angular-mocks` is cleaning up the `$rootScope` after each test, but doesn't take into account the fact that `$rootScope` might be mocked and not have a `$destroy()` method.

We need to fix this, but in the meantime you could work around the issue by providing a dummy `$destroy()` method on your mocked `$rootScope`. E.g.:

```js
module({
  $rootScope: {
    $destroy: angular.noop,
    $on: ...
  }
});
```

--
This still doesn't explain @darkmavis1980's problem with test failing even without a mocked `$rootScope` :confused:

**Comment by aciccarello** [comment]
Thanks for the update and suggestion @gkalpak. I was able to work around the issue.

**Comment by Narretz** [comment]
This is fixed since 1.5.2 iirc
